examples/cfd/stokes_mgn/pi_fine_tuning_gnn.py

```diff
@@ -0,0 +1,504 @@
+        self.nu = nu
+        self.dgl_graph = dgl_graph.to(self.device)
+        edge_tensor = torch.stack(
+            [dgl_graph.edges()[0], dgl_graph.edges()[1]], dim=1
+        ).to(self.device)
+        self.connectivity_tensor = compute_connectivity_tensor(
+            dgl_graph.nodes(), edge_tensor
+        )
+        self.connectivity_tensor = self.connectivity_tensor.to(self.device)
+
+        self.ref_u = torch.tensor(ref_u).float().to(self.device)
+        self.ref_v = torch.tensor(ref_v).float().to(self.device)
+        self.ref_p = torch.tensor(ref_p).float().to(self.device)
+
+        self.gnn_u = torch.tensor(gnn_u).float().to(self.device)
+        self.gnn_v = torch.tensor(gnn_v).float().to(self.device)
+        self.gnn_p = torch.tensor(gnn_p).float().to(self.device)
+
+        self.coords = torch.tensor(coords, requires_grad=True).float().to(self.device)
+        self.coords_inflow = (
+            torch.tensor(coords_inflow, requires_grad=True).float().to(self.device)
+        )
+        self.coords_noslip = (
+            torch.tensor(coords_noslip, requires_grad=True).float().to(self.device)
+        )
+
+        self.model = MeshGraphNet(
+            cfg.input_dim_nodes
+            + 128,  # additional 128 node features from fourier features
+            cfg.input_dim_edges,
+            cfg.output_dim,
+            aggregation=cfg.aggregation,
+            hidden_dim_node_encoder=cfg.hidden_dim_node_encoder,
+            hidden_dim_edge_encoder=cfg.hidden_dim_edge_encoder,
+            hidden_dim_node_decoder=cfg.hidden_dim_node_decoder,
+        ).to(self.device)
+
+        self.node_pde = Stokes(nu=self.nu, dim=2)
+
+        # note: this example uses the PhysicsInformer class from Modulus Sym to
+        # construct the computational graph. This allows you to leverage Modulus Sym's
+        # optimized derivative backend to compute the derivatives, along with other
+        # benefits like symbolic definition of PDEs and leveraging the PDEs from Modulus
+        # Sym's PDE module.
+
+        self.phy_informer = PhysicsInformer(
+            required_outputs=["continuity", "momentum_x", "momentum_y"],
+            equations=self.node_pde,
+            grad_method="least_squares",
+            device=self.device,
+            compute_connectivity=False,
+        )
+
+        self.optimizer = torch.optim.Adam(
+            self.model.parameters(),
+            lr=cfg.pi_lr,
+            fused=True if torch.cuda.is_available() else False,
+        )
+
+        self.scheduler = torch.optim.lr_scheduler.ExponentialLR(
+            self.optimizer, gamma=0.99935
+        )
+
+    def parabolic_inflow(self, y, U_max=0.3):
+        u = 4 * U_max * y * (0.4 - y) / (0.4**2)
+        v = torch.zeros_like(y)
+        return u, v
+
+    def loss(self):
+        out = self.model(
+            self.dgl_graph.ndata["x"], self.dgl_graph.edata["x"], self.dgl_graph
+        )
+
+        # inflow points
+        mask_inflow = (
```

Comment:
Is this the masking because we don't calculate the gradients on boundary elements/voxels/nodes? Should this be taken care of when the gradient is returned by the PhysicsInformer utility instead of the user having to code it up?

Reply:
So actually the `PhysicsInformer` utility will compute residuals on any points that the user passes. Since here, the points passed include the boundary points, the residuals computed on those points are not required as we need to apply BCs on them instead. Hence the masking was added. Potentially we have a higher level of abstraction at a later point?